src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -0,0 +1,131 @@
+                if (reader._stringHasEscaping)
+                {
+                    int idx = propertyName.IndexOf(JsonConstants.BackSlash);
+                    Debug.Assert(idx != -1);
+                    propertyName = GetUnescapedString(propertyName, idx);
+                }
+
+                state.Current.JsonPropertyInfo = state.Current.JsonClassInfo.GetProperty(options, propertyName, ref state.Current);
+                if (state.Current.JsonPropertyInfo == null)
+                {
+                    if (state.Current.JsonClassInfo.DataExtensionProperty == null)
+                    {
+                        state.Current.JsonPropertyInfo = JsonPropertyInfo.s_missingProperty;
+                    }
+                    else
+                    {
+                        ProcessMissingProperty(propertyName, options, ref reader, ref state);
+                    }
+                }
+                else
+                {
+                    state.Current.PropertyIndex++;
+                }
+            }
+        }
+
+        private static void ProcessMissingProperty(
+            ReadOnlySpan<byte> unescapedPropertyName,
+            JsonSerializerOptions options,
+            ref Utf8JsonReader reader,
+            ref ReadStack state)
+        {
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonClassInfo.DataExtensionProperty;
+
+            Debug.Assert(jsonPropertyInfo != null);
+            Debug.Assert(state.Current.ReturnValue != null);
+
+            IDictionary extensionData = (IDictionary)jsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue);
+            if (extensionData == null)
+            {
+                Type type = jsonPropertyInfo.DeclaredPropertyType;
+
+                // Create the appropriate dictionary type. We already verified the types.
+                Debug.Assert(type.IsGenericType);
+                Debug.Assert(type.GetGenericArguments().Length == 2);
+                Debug.Assert(type.GetGenericArguments()[0].UnderlyingSystemType == typeof(string));
+                Debug.Assert(
+                    type.GetGenericArguments()[1].UnderlyingSystemType == typeof(object) ||
+                    type.GetGenericArguments()[1].UnderlyingSystemType == typeof(JsonElement));
+
+                extensionData = (IDictionary)options.GetOrAddClass(type).CreateObject();
+                jsonPropertyInfo.SetValueAsObject(state.Current.ReturnValue, extensionData);
+            }
+
+            JsonElement jsonElement;
+            using (JsonDocument jsonDocument = JsonDocument.ParseValue(ref reader))
+            {
+                jsonElement = jsonDocument.RootElement.Clone();
+            }
+
+            string keyName = Encoding.UTF8.GetString(unescapedPropertyName);
+
+            // Currently we don't apply any naming policy. If we do, we'd have to pass it onto the JsonDocument.
```

Comment:
So there are two different things that can have a naming policy applied to it.
1. The dictionary key
2. The JSON content

Json.NET has an extension point for changing the dictionary key. On the other hand the Json.NET serializer doesn't modify the extension data content. An approach I took with the serializer is JToken (i.e. JsonElement) is already JSON, and the serializer always writes it unchanged.

Reply:
> On the other hand the Json.NET serializer doesn't modify the extension data content. An approach I took with the serializer is JToken (i.e. JsonElement) is already JSON, and the serializer always writes it unchanged.

I am fine with this and it would remove the need to pass in the serializer options to the `JsonDocument`.

@rynowak, @KrzysztofCwalina - what do you think? Is it necessary for the "overflow" data to honor the serializer options?

Reply:
Here's the settings not on the reader\writer but could be applied:
- Naming on serialize (not in reader options, plus would be a new option on JsonSerializerOptions)
- IgnoreNulls (not in reader options)

Here's the settings on reader\writer options that could applied (they should be applied IMO but currently aren't):
- AllowTrailingCommas
- MaxDepth
- ReadCommentHandling
- WriteIndented

Reply:
For now, pending future discussion, I will pass in the JsonReaderOptions\JsonWriterOptions like we do for other calls into the reader\writer. I will not add any support for property naming or IgnoreNull at this time.